Allow to set the version of the unittest plugin

camalot opened the pull request:
accept input for `unittest-version`

if not set, it will use the default "latest" version.
if it is set, specify that version during the install of the plugin

camalot committed:
feat: add input to set the unittest plugin version

Add an action input to set the version of helm-unittest to use. This is
particularly useful in times of instability and breaking changes.

d3adb5 commented:
Thank you for working on this, @camalot! Could you add a test case in the CI workflow for this, however? I understand there is incompatibility between versions 0.2.x and 0.3.x of helm-unittest, but maybe as a test case we could try installing 0.3.3 and checking that that is the version that was installed?

I'm thinking we could always let the action itself fail, as is done in existing test scenarios, and just check the version afterwards. What do you think?

camalot commented:
Sure thing. I will look into that. I had to throw a quick fix together today as a lot of our repos started failing the other day, and i just tracked it down today to being the most recent build of `helm-unittest` (related issue: https://github.com/helm-unittest/helm-unittest/issues/219)

we have been using this action up until when `v0.3.5` was released without issue. 

I will probably have some time tomorrow to take a look at adding tests around it.

camalot commented:
ok, so i had a few minutes...

here is what I was thinking after looking at the existing test cases.

``` yaml
  test-action-unittest-version:
    name: Unittest Version
    runs-on: ubuntu-latest
    timeout-minutes: 10

    steps:
      - name: Checkout repository files
        uses: actions/checkout@v3
      - name: Run composite action
        uses: ./
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          unittest-version: v0.3.4
          
      - name: Check that helm-unittest is installed
        run: helm plugin list | sed 1d | awk '{print $1}' | grep -q '^unittest$'

      - name: Check that helm-unittest version is v0.3.4
        run: helm plugin list | sed 1d | awk '{print $2}' | grep -q '^v0.3.4$'
```

verify that it is installed, and the specified version is installed.

d3adb5 commented:
That looks good to me, feel free to commit it. If it doesn't run automatically, I'll approve the workflow here so we can see how it behaves. Thanks, @camalot!

d3adb5 commented:
Ah, Helm doesn't prefix versions in the plugin table with `v`, that's why it failed. The fix is simple, luckily.

camalot commented:
yeah, i also tweaked it to include the name of the plugin, just in case. didnt want to have it "pass" if there was some other plugin that had the version.

d3adb5 commented on .github/workflows/ci.yml:
```diff
@@ -124,6 +124,28 @@ jobs:
+      - name: Check that helm-unittest is installed
+        run: helm plugin list | sed 1d | awk '{print $1}' | grep -q '^unittest$'
+
+      - name: Check that helm-unittest version is v0.3.4
+        run: helm plugin list | sed 1d | awk -v OFS='\t' '{print $1, $2}' | grep $'unittest\t0.3.4$'
```
This `grep` call bothers me a little. It's already iffy that we're relying on the output of `helm list` being a table formatted the way it is currently. What to say of it trying to accommodate other plugins with longer names?

Instead, I propose the following solution:

```suggestion
      - name: Check that helm-unittest version is v0.3.4
        run: helm plugin list | awk '/^unittest\s/ { found = 1; print $2 } END { exit !found }' | grep -q '^0.3.4$'
```

We can also use this bit of awk for the other checks. I'm assuming here that GitHub Actions sets `pipefail`.

d3adb5 commented on action.yml:
```diff
@@ -59,8 +64,14 @@ runs:
-
-        helm plugin install https://github.com/helm-unittest/helm-unittest >/dev/null 2>/dev/null
+        helm_ut_plugin_version="${{ inputs.unittest-version }}"
+        if [[ ! -z "${helm_ut_plugin_version// }" ]]; then
```
In keeping with POSIX compliance:

```suggestion
        if [ -n "$helm_ut_plugin_version" ]; then
```

We do lose the ability to "trim" the input, but I think that's okay since the user has to go out of their way to send just whitespace.

d3adb5 reviewed (changes requested):
Left two suggestions, but otherwise looks good to me! Would you be so kind as to squash your commits and follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/), @camalot?

camalot commented:
I think i squashed correctly. its been awhile. 

and I think the commit message follows the Conventional Commits.

camalot commented:
I should probably update the readme as well...

I will get to that tomorrow.

d3adb5 committed:
docs: add info about unittest-version input to readme

Add information about the unittest-version action input to the project's
README file.

d3adb5 committed:
style: remove trailing whitespace & extra empty lines

Remove trailing whitespace and extra empty lines from the action file.

d3adb5 commented:
I took the liberty of updating the README and removing trailing whitespace. Thank you, @camalot! I'm merging this for now and creating a new release.
